pkg/apis/pipeline/v1alpha1/verificationpolicy_validation_test.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2022 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/tektoncd/pipeline/test/diff"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+)
+
+func TestVerificationPolicy_Invalid(t *testing.T) {
+	tests := []struct {
+		name               string
+		verificationPolicy *v1alpha1.VerificationPolicy
+		want               *apis.FieldError
+	}{{
+		name: "missing Resources",
+		verificationPolicy: &v1alpha1.VerificationPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "vp",
+			},
+			Spec: v1alpha1.VerificationPolicySpec{
+				Authorities: []v1alpha1.Authority{
+					{
+						Name: "foo",
+						Key: &v1alpha1.KeyRef{
+							Data: "inline_key",
+						},
+					},
+				},
+			},
+		},
+		want: apis.ErrMissingField("resources"),
+	}, {
+		name: "invalid Resources",
+		verificationPolicy: &v1alpha1.VerificationPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "vp",
+			},
+			Spec: v1alpha1.VerificationPolicySpec{
+				Resources: []v1alpha1.ResourcePattern{{"^["}},
+				Authorities: []v1alpha1.Authority{
+					{
+						Name: "foo",
+						Key: &v1alpha1.KeyRef{
+							Data: "inline_key",
+						},
+					},
+				},
+			},
+		},
+		want: apis.ErrInvalidValue("^[", "ResourcePattern", fmt.Sprintf("ResourcePattern cannot be compiled by regex: error parsing regexp: missing closing ]: `[`")),
+	}, {
+		name: "missing Authoritities",
+		verificationPolicy: &v1alpha1.VerificationPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "vp",
+			},
+			Spec: v1alpha1.VerificationPolicySpec{
+				Resources: []v1alpha1.ResourcePattern{{".*"}},
+			},
+		},
+		want: apis.ErrMissingField("authorities"),
+	}, {
+		name: "missing Authority key",
+		verificationPolicy: &v1alpha1.VerificationPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "vp",
+			},
+			Spec: v1alpha1.VerificationPolicySpec{
+				Resources: []v1alpha1.ResourcePattern{{".*"}},
+				Authorities: []v1alpha1.Authority{
+					{
+						Name: "foo",
+						Key:  &v1alpha1.KeyRef{},
+					},
+				},
+			},
+		},
+		want: apis.ErrMissingOneOf("key[0].data", "key[0].secretref"),
+	}, {
+		name: "has both data and secretref",
```

Comment:
name the case by what it is testing (what is missing), not by what it has -- this helps me understand what's expected

Reply:
thanks! "has both data and secretref" is an invalid case, how about renaming it to "should not have both data and secretref" ?